Feature/tiling demo

mattdelsordo opened the pull request:
## Overview
This PR plugs mapnik into the database in order to generate map tiles, which then get served up through an API.


### Demo
[Example deployment (27 Jun 2018)](https://d28f32z0lbuk2f.cloudfront.net/tile/13/2385/3103.png)
![image](https://user-images.githubusercontent.com/16990679/41936725-21c71420-795c-11e8-802f-818185f24b87.png)
![image](https://user-images.githubusercontent.com/16990679/41936741-2e65a052-795c-11e8-87c3-1a0f0fc53109.png)


### Notes
Current development environment setup doesn't necessarily obey the quirks of the real API Gateway, tiles are served from what is basically an express server. Also, the `pwd_inlets` data set is very large on a per-tile basis which may cause the lambdas to time out. Increase the value of `TIMEOUT` in `.env` if you are consistently getting un-viewable images or 502 errors.

#### Accept headers
API Gateway currently has issues serving images that can be simply displayed in-browser, so you have to manually add an `Accept:image/png` header to your lambda requests to receive proper images. This can be done either with curl (`-H "Accept:image/png"`) or in production by setting up a proxy (e.g. with CloudFront) that adds the header to all lambda requests.


## Testing Instructions
 * Run `./scripts/update` and `./scripts/server` to run the local tileserver.
 * Open `src/demo/pwd_inlets_index.html` to generate and display tiles through Leaflet.
 * Run `./scripts/publish` (optional `--new` flag if you haven't deployed the project before). Assuming your credentials are set up properly it should deploy within about a minute.
 * Change the value of TILESERVER_PATH to load files the lambdas you just deployed.

Resolves #10, resolves #19, resolves #31, resolves #32

mattdelsordo committed:
Initial commit of nonfunctional skeleton code

mattdelsordo committed:
Add mapnik hello world example to explore intended mapnik functionality

mattdelsordo committed:
Establish connection with PostGIS

mattdelsordo committed:
Enable writing tiles to file to test rendering

mattdelsordo committed:
Disable unneeded logging

mattdelsordo committed:
Add demo for pa_gardens tile rendering

mattdelsordo committed:
Remove reliance on custom bounding box conversions

mattdelsordo committed:
Refactor _mapnik-hello-world functionality into existing project structure

Recycled code from _mapnik-hello-world.js into code structure of
the lambnik API. grid and image functions now return Promises to
handle asynchronous rendering behavior.

mattdelsordo committed:
Clean up vestigial testing and configuration options

mattdelsordo committed:
Fix issues that prevented lambdas from deploying

mattdelsordo committed:
Remove local volume from PostGIS configuration

Database files were being shared with the host which complicated
clean installs of the local environment.

mmcfarland commented on node_lambnik/docker-compose.yml:
```diff
@@ -14,8 +14,6 @@ services:
     expose:
       - "5432"
-    volumes:
-      - /opt/dbdata/lambnik:/var/lib/postgresql/data
```
I think the reason this is normally included is to persist the database between the cycles of `docker-compose` up/down.  I'd verify that this change still works for your normal dev workflow.

KlaasH replied:
I suggested this because we were seeing differences between our ostensibly identical dev instances. The database here is basically a test fixture, so I think making it more ephemeral makes sense. Persisting it on the host seems to me like it adds more potential gotchas than benefits.

mattdelsordo replied:
My stance on having that volume is that it's more of a hassle to have that data persisted and then be confused when you forget about it and need freshly create the environment than it is to have to run `update` more often, since the `update` --> `server` process is already specified in the README but it might not be totally clear that the database info is being persisted locally.

KlaasH commented:
This currently has demo files for both `pa_gardens` and `pwd_inlets`.  I think it would make sense to pick one to get working and leave the other out of the repo for now.  My feeling is that using `pa_gardens`, since it's simpler, is preferable for the initial demo.  That would also mean we could remove the `inlet` filtering from node-lambnik, which I think makes sense since that's very specific to the original Lambnik demo and might not match where we ultimately want to go with filtering config.

KlaasH commented on node_lambnik/src/demo/pa_gardens_index.html:
```diff
@@ -0,0 +1,58 @@
+
+
+        // add garden map
+        L.tileLayer('http://localhost:9001/tile/{z}/{x}/{y}.png').addTo(map)
```
Worth using a variable here, like in the original Lambnik demo, to make it easy to switch between the local and remote tilesource.

KlaasH commented on node_lambnik/src/tiler/package.json:
```diff
@@ -43,6 +44,8 @@
   },
   "dependencies": {
+    "@mapbox/sphericalmercator": "^1.1.0",
+    "babel-runtime": "6",
```
Am I remembering correctly that you weren't sure if this was necessary?  Or am I confusing this with a conversation about something else?  In any case, where does this come into play?

mattdelsordo replied:
I got SphericalMercator working, turns out I was just telling it to use the wrong projection. Re: babel-runtime, I ran into an issue with getting Promises to run reliably after transpilation, so I added that at the internet's suggestion and it solved whatever issue that was.

KlaasH commented on node_lambnik/src/tiler/src/tile-interface.js:
```diff
@@ -0,0 +1,65 @@
+import { image, grid } from './tiler'
+
+
+// TODO: flesh out how errors are handled
```
On Urban Apps, we recently had some discussions about TODO items and arrived at a process whereby any TODO added to the code (i.e. anything you're planning to merge, not just a note to be handled before finalizing a PR) should have a corresponding issue in the repo, with the issue number included in the TODO comment.  I think it's a good system, and that these TODO items are worth adding issues for.

mattdelsordo replied:
Sounds good, I'll add those in.

mattdelsordo committed:
Remove unneeded demo materials

KlaasH reviewed (commented):
Lots of good stuff here.  The image encoding thing (whatever it is) is certainly a hassle.  Having said before that it would make sense to merge this and then deal with that part, I'm now feeling like getting the basic demo page working, both locally and remotely, is probably still a better place to stop.  I had been thinking this was the `Accept` header issue, but that should be handled by the manually configured CloudFront proxy.

What's here looks good, but I think it makes sense to prune it down to 1 demo (i.e. only `pa_gardens`, as noted above) and get it to the point of working locally and remotely before closing this out.

mattdelsordo committed:
Properly serve images on the local server

Development server uses a locally modified claudia-local-api that
solves it's inability to serve binary data. PR these changes if
possible and switch back to the node module if/when the changes
make it there.

mattdelsordo committed:
Add variable tileserver path to demo

mattdelsordo committed:
Remove writing to file now that image serving is confirmed working

mattdelsordo committed:
Add themed icon/move resources to their own directory

mattdelsordo committed:
Move binary content serving test to tile-interface

mattdelsordo committed:
Refactor tiler code to operate asynchronously

mattdelsordo committed:
Update usage information and add usage to landing endpoint

mattdelsordo committed:
Re-throw errors so that they actually get printed/returned to client

mattdelsordo committed:
Switch back to pwd_inlets data set because it is what currently exists in production

mattdelsordo committed:
Correct usage instructions

mattdelsordo committed:
Simplify map style for speed reasons

mattdelsordo committed:
Add env to simplify switching between data sets

mattdelsordo committed:
Add testing script

mattdelsordo committed:
Make claudia-local-api a git-based dependency rather than manually including the files

mattdelsordo committed:
Add CloudFront info to README

mattdelsordo committed:
Update production URL in demo

KlaasH commented on node_lambnik/src/tiler/src/api.js:
```diff
@@ -3,40 +3,47 @@
 
-api.get('/', () => 'Hello, world!')
+// Add pngs to binary media types
+//api.setBinaryMediaTypes(['image/png'])
```
Looks like this was something that was tried but didn't prove useful.  Should just remove it.

KlaasH commented on node_lambnik/src/tiler/src/path-config.js:
```diff
@@ -0,0 +1,9 @@
+export const TILE_PATH = '/tile/{z}/{x}/{y}'
+export const GRID_PATH = '/grid/{z}/{x}/{y}'
+export const FAVICON_PATH = '/favicon.ico'
+export const IMAGE_TEST_PATH = '/img'
```
Looks like this was made a separate file so it can be used in both `api.js` and `tiler-interface.js`.  I think when you do #37 it should be moved back inline.  Either in the actual `api.get()` calls or as a config object at the top of `api.js`.  I'm currently leaning toward the former.

KlaasH commented on node_lambnik/src/tiler/src/tiler.js:
```diff
@@ -0,0 +1,153 @@
+                })
+            })
+        }))
+        .then(result => result)
```
This seems odd.  Does it do something useful?

KlaasH commented on node_lambnik/src/tiler/src/tiler.js:
```diff
@@ -0,0 +1,153 @@
+    const grd = new mapnik.Grid(TILE_WIDTH, TILE_HEIGHT)
+
+    return createMap(z, x, y)
+        .then((map) => new Promise((resolve, reject) => {
```
~You're making a new promise here but not calling the `resolve` or `reject` functions.~
Actually, I see this is in the `grid` function and is addressed in PR #38, so that's fine.

KlaasH commented on node_lambnik/env-template:
```diff
@@ -8,3 +8,9 @@ TILEGARDEN_DB=
+DB_TABLE=
+GEOMETRY_FIELD=geom
+EPSG=4326
+TIMEOUT=5
```
Probably makes sense to namespace these.  E.g. `TILEGARDEN_DB_TABLE`, `TILEGARDEN_GEOMETRY_FIELD`, `TILEGARDEN_GEOMETRY_EPSG`.  And a more descriptive name for what timeout that's setting.

mattdelsordo replied:
I think I'm going to address this in the CartoCSS PR where I'm already refactoring a lot of the configuration setup.

KlaasH commented on node_lambnik/scripts/test:
```diff
@@ -0,0 +1,38 @@
+        main $SERVER_COMMAND
+    fi
+    exit
+fi
```
I think it might be better to leave this out of this PR and save it for issue #14.  Whether it stays or goes, I'd rather that it clean up after itself than create a `test-results` directory that sticks around.

mattdelsordo replied:
I'll remove it, all the scripts do are try to download the images from the tile server and put them in that directory, it's sort of vestigial from before the binary content situation was sorted out.

KlaasH reviewed (approved):
Works well and looks good.  A few fixups and questions to consider, but :+1: with whatever you decide to do about them.

mattdelsordo committed:
Clean up vestigial code
